File: python/lrucache.py
class LinkedList:
    def __init__(self, key, value):
        self.next = None
        self.prev = None
        self.key = key
        self.value = value
        

class LRUCache:
    def __init__(self, maxsize):
        self.maxsize = maxsize or 1
        self.currsize = 0
        self.cache = {}
        self.head = None
        self.tail = None
  
    # Time -> 0(1) Space -> 0(1)
    def insertKeyValuePair(self, key, value):
        if key in self.cache:
            pointer = self.cache[key]
            pointer.value = value
            return
        if self.currsize == self.maxsize:
            self.deleteKey()
        pointer = LinkedList(key=key, value=value)
        self.cache[key] = pointer
        self.currsize = min(self.currsize+1, self.maxsize)
        if self.head == None:
            self.head = pointer
            self.tail = pointer
            return
        pointer.prev = self.tail
        self.tail.next = pointer
        self.tail = pointer
        
    # Time -> 0(1) Space -> 0(1)   
    def getValueFromKey(self, key):
        if key in self.cache:
            pointer = self.cache[key]
            self.updateRecentlyUsedKey(pointer)
            return pointer.value
        return None
    
    # Time -> 0(1) Space -> 0(1)
    def getMostRecentKey(self):
        if self.currsize == 0:
            return None
        return self.tail.key
    
    
    def updateRecentlyUsedKey(self, pointer):
        if pointer == self.tail:
            return
        elif pointer == self.head:
            self.head = self.head.next
            self.head.prev = None
        else:
            pointer.next.prev = pointer.prev
            pointer.prev.next = pointer.next
            
        pointer.prev = self.tail
        self.tail.next = pointer
        self.tail = pointer
        self.tail.next = None
        
    
    def deleteKey(self):
        key = self.head.key
        self.head = self.head.next
        if self.head != None:
            self.head.prev = None
        self.cache.pop(key)
        
        
lru = LRUCache(maxsize=4)
lru.insertKeyValuePair("a", 1)
lru.insertKeyValuePair("b", 2)
print(lru.getMostRecentKey())
print(lru.getValueFromKey("a"))
lru.insertKeyValuePair("c", 3)
print(lru.getValueFromKey("d"))
lru.insertKeyValuePair("d", 4)
lru.getValueFromKey("b")

File: go/LRUCache/lrucache.go
package main

import (
	"fmt"
)

type LinkedList struct {
	value int
	key string

	Prev *LinkedList
	Next *LinkedList
}


type LRUCache struct {
	maxSize int
	currSize int

	cache 	map[string]*LinkedList
	head	*LinkedList
	tail 	*LinkedList
}

func NewLRUCache(size int) *LRUCache{
	return &LRUCache{maxSize: size, cache: make(map[string]*LinkedList, 0)}
}


// Time -> 0(1) Space -> 0(1)
func (lru *LRUCache) InsertKeyValuePair(key string, value int) {
	if pointer, exists := lru.cache[key]; exists {
		pointer.value = value
		lru.NewRecentlyUsed(pointer)
		return
	}
	if lru.currSize == lru.maxSize {lru.RemoveLRU(lru.head.key)}
	pointer := &LinkedList{value: value, key: key}
	lru.cache[key] = pointer
	lru.currSize = min(lru.currSize+1, lru.maxSize)
	if lru.head == nil {
		lru.head = pointer
		lru.tail = pointer
		return
	}
	pointer.Prev = lru.tail
	lru.tail.Next = pointer
	lru.tail = pointer
}


// Time -> 0(1) Space -> 0(1)
func (lru *LRUCache) GetValueFromKey(key string) (int, bool) {
	if val, exists := lru.cache[key]; exists {
		lru.NewRecentlyUsed(val)
		return val.value, true
	}
	return -1, false
}


// Time -> 0(1) Space -> 0(1)
func (lru *LRUCache) GetMostRecentKey() (string, bool) {
	if lru.currSize == 0 {
		return "", false
	}
	return lru.tail.key, true
}


func (lru *LRUCache) RemoveLRU(key string) {
	if lru.head.Next == nil {
		lru.head = nil
	}else {
		lru.head = lru.head.Next
		lru.head.Prev = nil
	}
	delete(lru.cache, key)
}


func(lru *LRUCache) NewRecentlyUsed(pointer *LinkedList) {
	if pointer == lru.tail {return}
	if pointer == lru.head {
		lru.head = lru.head.Next
		lru.head.Prev = nil

	}else {
		pointer.Prev.Next = pointer.Next
		pointer.Next.Prev = pointer.Prev
	}
	pointer.Prev = lru.tail
	lru.tail.Next = pointer
	lru.tail = pointer
	lru.tail.Next = nil
}

func min(x, y int) int {
	if x < y {return x}
	return y
}

func main() {
	lru := NewLRUCache(3)
	lru.InsertKeyValuePair("a", 1)
	lru.InsertKeyValuePair("b", 2)
	fmt.Println(lru.GetMostRecentKey())
	fmt.Println(lru.GetValueFromKey("a"))
	lru.InsertKeyValuePair("c", 3)
	fmt.Println(lru.GetMostRecentKey())
	lru.InsertKeyValuePair("d", 4)
	fmt.Println(lru.GetValueFromKey("b"))
}